### ISSUE TITLE
Add Type Registry convenience methods

### ISSUE THREAD

**Issue by dmitrizagidulin** [solid/solid.js#32]
Add client-side support for the [type registry](https://github.com/solid/solid/blob/master/proposals/data-discovery.md) (allow an app to register a given type as stored in an instance or a container)

- [x] Add parsing of Private and Public type registry links
- [x] Add the loading of the Preferences file to `Solid.identity.loadExtendedProfile()` (which currently only loads the `seeAlso` and `sameAs` to be parsed). Which then enables:
- [x] Provide convenience methods to actually load & parse the public and private typeIndexes (at the moment, the profile just parses the links)
- [x] Differentiate between the parsed public and private typeIndexes (since the user/app dev/UI needs to tell them apart, when asking where to register what).
- [x] Add methods to register a type
- [x] Add methods to de-register types

**Comment by dmitrizagidulin** [comment]
To do:

- [ ] Add the loading of the Preferences file to `Solid.identity.loadExtendedProfile()` (which currently only loads the `seeAlso` and `sameAs` to be parsed). Which then enables:
- [ ] Parse the link to the Private type registry index from the Preferences file
- [ ] Provide convenience methods to actually load & parse the public and private typeIndexes (at the moment, the profile just parses the links)
- [ ] Differentiate between the public and private typeIndexes (since the user/app dev/UI needs to tell them apart, when asking where to register what). Tim currently does this by matching on the source:

  ```
context.index.private = kb.each(me, ns.solid('typeIndex'), undefined, context.preferencesFile);
context.index.public = kb.each(me, ns.solid('typeIndex'), undefined, context.publicProfile);
  ```

**Comment by dmitrizagidulin** [comment]
Updated the issue description with to-do steps.

**Comment by dmitrizagidulin** [comment]
To Do:

Implement the function:

```javascript
/**
 * @param profile {SolidProfile} Loaded WebID profile
 * @param rdfClass {rdf.NamedNode} Type to register in the index
 * @param [accessLevel='private'] {String} Optional type registry access level
 *   (whether to register in a private or public index).
 *   One of 'public' or 'private', defaults to 'private'.
 */
function registerType (profile, rdfClass, accessLevel) {
}
```

The workflow will go, roughly:

**Initially:** Check if the profile has a link to the corresponding private or public type registry index.

**Type Registry Index link exists:** Check to see if the type is already registered there. If yes, do a PATCH to update it. If not, do a PATCH to insert it.

**Type Registry Index links not in profile:** in this case, check to see if the profile contains a link to the Settings container (the one which has the Preferences file). If settings container doesn't exist, create it (see below), and then:

  - Create a `/settings/publicTypeIndex.ttl` resource with a Public ACL
  - Create a `/settings/privateTypeIndex.ttl` resource (default private ACL)
  - PATCH-insert links to those two index resources into the WebID profile resource

**Settings container doesn't exist:** If there's no link to the Preferences file in the profile, check to see if the user has a link to a Root Storage container, and then do the following:

  - Create the Default Container `/settings/`, with the default private ACL
  - Create an empty `/settings/prefs.ttl` resource
  - Link to `pres.ttl` from the WebID profile (via PATCH)
  - Proceed to create & link the registry index files as described above

**No link to Root Storage:** If there's no link to root storage in the WebID profile, flip the tables, and throw an `Invalid Profile - No Storage link detected` exception.

**Comment by deiu** [comment]
I wonder if we can just get the parent container of the prefs.ttl document (if it exists), or otherwise the profile document and use that the base for the typeIndex documents.

```
dirname('/settings/prefs.ttl') -> '/settings/'
```
or
```
dirname('/profile/card') -> '/profile/'
```

**Comment by dmitrizagidulin** [comment]
Yea, getting the parent container for the prefs.ttl file is implied in the above.

**Comment by dmitrizagidulin** [comment]
I'm not sure about using the `/profile/` container, however. (Since that one's public by default).

**Comment by deiu** [comment]
What I meant is that we don't have to create a `/settings/` container. Just go with whatever setup the user currently has.

**Comment by dmitrizagidulin** [comment]
Check, on the same page.

**Comment by dmitrizagidulin** [comment]
(Updated the function spec to include the actual location where the class is being registered to, whoops)

**Comment by dmitrizagidulin** [comment]
One of the implications here is that there should be only one private type index, and only one public type index (otherwise, the client doesn't have a good way of choosing *which* private index to register something to).  - opened issue https://github.com/solid/solid.js/issues/60

**Comment by dmitrizagidulin** [comment]
After extended conversations, decided not to fix the user's profile in the `profile.registerType()` method (create the Preferences file if missing, create the registry indexes if missing, etc). Will leave that to a separate feature.

**Comment by dmitrizagidulin** [comment]
`profile.registerType()` implemented.
Remaining todo: implement `profile.unregisterType()`.

**Comment by dmitrizagidulin** [comment]
`profile.unregisterType()` implemented in PR #69.
Closing.
